asdf/extension/_compressor.py

```diff
@@ -0,0 +1,84 @@
+"""
+Compressor is an interface to implement extensions to the compression
+module. Extensions will typically subclass the Compressor ABC and
+provide that subclass as a setuptools entry point.  Decompressor
+is likewise used for decompression.
+
+Note that this interface has similar patterns to Converter.  This
+interface is designed for compression and decompression of ASDF
+binary array blocks, while Converter is designed for serialization
+of custom Python types into the YAML tree.
+"""
+
+import abc
+
+class Compressor(abc.ABC):
+    """
+    Abstract base class for plugins that compress binary data.
+
+    Implementing classes must provide the `labels` property, and
+    the `compress()` method.
+    May also provide a constructor that takes kwargs that are
+    set via `asdf.get_config().compression_options[label] = kwargs`.
+    """
+    @classmethod
+    def __subclasshook__(cls, C):
+        if cls is Compressor:
+            return (hasattr(C, "labels") and
+                    hasattr(C, "compress"))
+        return NotImplemented # pragma: no cover
+
+
+    @abc.abstractproperty
+    def labels(self):
+        """
+        Get the string labels that this Compressor
+        is able to compress.
+
+        Returns
+        -------
+        iterable of str
+            str labels handled by this class
+        """
+        pass # pragma: no cover
+
+
+    @abc.abstractmethod
+    def compress(self, data):
+        """
+        Returns compressed data.
+        """
+        pass # pragma: no cover
+
+
+class Decompressor(abc.ABC):
+    """
+    Abstract base class for plugins that decompress binary data.
+
+    Implementing classes must provide the `labels` property,
+    and the `decompress()` or `decompress_into()`
+    methods.  May also provide a constructor that takes kwargs that are
+    set via `asdf.get_config().decompression_options[label] = kwargs`.
+    """
+    @classmethod
+    def __subclasshook__(cls, C):
+        # TODO: is this the right way to enforce that a subclass define `decompress()` or `decompress_into()`?
+        if cls is Compressor:
+            return (hasattr(C, "labels") and
+                    ( hasattr(C, "decompress") or
```

Comment:
Another option (instead of separate decompress/decompress_into methods) is to mimic the [numcodecs interface](https://numcodecs.readthedocs.io/en/latest/abc.html#numcodecs.abc.Codec.decode) and make the destination array an optional keyword argument to decompress.  That has the advantage of allowing us to make decompress an `@abc.abstractmethod`.

Reply:
I did think about that, but the presence of `decompress_into` lets the `decompress` function know whether to provide an `out` arg and whether the return value is bytes to be concatenated or a number of bytes written.  It's possible to require that compressors handle the `out` arg with simple concatenation if they don't natively support writing to a user-supplied array, but then every implementation needs to repeat that logic.  I guess we could detect whether the return type is an integer and assume successful writing in-place if so.  Another option would be a field like `Decompressor.supports_append`.

Reply:
This may be another symptom of my more general misunderstanding of how these methods will be used, but does it work to always return the decompressed array?  Something like this:

```python
def decompress(self, data, out=None):
  """
  Decompress block data.
  
  Parameters
  ----------
  data : np.ndarray
    Compressed bytes.
  out : np.ndarray, optional
    Optional pre-allocated output array.  If present, must be exactly the size required
    to store the decompressed data.

  Returns
  -------
  np.ndarray
    Decompressed data.  This will be the same array object as `out` if that parameter
    was provided.
  """
  pass
```
Then if an implementation uses a compressor that doesn't natively support writing to a user-supplied array, it just has to do something like this:
```python
def decompress(self, data, out=None):
  uncooperative_compressor = UncooperativeCompressor()
  result = uncooperative_compressor.decompress(data)
  if out is not None:
    out[:] = result
    return out
  else:
    return result
```

Reply:
That almost works, but you need a way to return the number of decompressed bytes, because you don't know the size of the decompressed block.  All you know is the total size of all decompressed blocks, because that's the logical array size that ASDF tracks in the tree.  One way to return the decompressed size is to advance the `out` "pointer", e.g. `return out[len(result):]`; another is to just return an int.  So it can work, it just requires that each decompressor implement this logic.

Unfortunately, this would also require wrapping the `zlib` and `bz2` built-in compressors.  Currently, the API is simple enough that it's compatible with `zlib.decompressobj` and `bz2.BZ2Decompressor`, but neither of them support the `out` parameter.

Reply:
The ASDF block header actually does include the number of decompressed bytes in a field, see `data_size` here:

https://asdf-standard.readthedocs.io/en/1.5.0/file_layout.html#block-header

I'd like to wrap zlib and bz2 in this interface anyway, so that everything is nice an uniform.  I don't see that as much of a downside since the wrapper will be thin.

Reply:
Sorry, I was referring to compression blocks, not ASDF blocks.  There can be multiple compression blocks per ASDF block (compression blocks are the ones sized to `DEFAULT_BLOCK_SIZE = 1 << 22` in `compression.py`), and the decompressed size of each of those isn't stored anywhere.  All you have is the total size (the `data_size`), which you can use to pre-allocate an array that the decompressed compression blocks can fill.

But if we're wrapping the built-in compressions anyway, then I think the int-return scheme is perfectly acceptable, as long as we document this API for extension implementers.

Reply:
Oh, ha, I didn't even know we did that.  Does that seem strange to you, that ASDF is only compressing 4 MB at a time?  I would think that we'd want to let the individual compression algorithms tune themselves.  The LZ4 compressor is writing out a header with each block, which sounds more like N compressed files cat'ed together?

Reply:
It's a little strange, but there are some pros, I think.  A pragmatic one is that in the case of Blosc and lz4 (at least the `lz4.block` API), there are 32-bit size limits that require some external blocking to deal with arrays larger than 2 GB.  But even in the absence of that problem, I think blocking enables efficient overlap of decompression and IO in a very general way.  One thread can read blocks and pass them to a decompression thread, which can always decompress the data because it is completely self-contained within that 4 MB.  The compression doesn't need to have any special "streaming" capability.  It's obviously not something being used right now, but might be very nifty in the future.

Reply:
My sentiment is that we don't want to miss out on possible optimizations that the underlying compression library may implement (screaming fast parallel compression in a C extension, for example).  IMO it's worth trading additional complexity in the `asdf.extension.Compressor` implementations to get that.  @perrygreenfield and @jdavies-st what do you think?

Reply:
This PR does allow tuning of the compression block size via the compression options mechanism we've been working on.  Maybe a value of `compression_block_size=None` or `NO_BLOCK` can force the use of one block?